lib/charms/loki_k8s/v1/loki_push_api.py

```diff
@@ -2313,6 +2382,123 @@ def _containers(self) -> Dict[str, Container]:
+        self.disable()
+
+    def enable(self):
+        """Enable log forwarding."""
+        self.handle_logging(enabled=True)
+
+    def disable(self):
+        """Disable log forwarding."""
+        self.handle_logging(enabled=False)
+
+    def _build_log_target(self, endpoint, index, enabled=False):
+        """Build a log target for the log forwarding Pebble layer."""
+        dest_name = f"loki{index}"
+        services_value = ["all"] if enabled else ["-all"]
+
+        return {
+            dest_name: {
+                "override": "merge",
+                "type": "loki",
+                "location": endpoint,
+                "services": services_value,
+                "labels": {
+                    "product": "Juju",
+                    "charm": self.topology._charm_name,
+                    "juju_model": self.topology._model,
+                    "juju_model_uuid": self.topology._model_uuid,
+                    "juju_application": self.topology._application,
+                    "juju_unit": self.topology._unit,
+                },
+            }
+        }
+
+    def _build_log_targets(self, endpoints: Optional[List[str]], enable=False):
+        """Build the targets for the log forwarding Pebble layer."""
+        targets = {}
+        if endpoints:
+            for i, endpoint in enumerate(endpoints):
+                targets.update(self._build_log_target(endpoint, i, enable))
+        return targets
+
+    def handle_logging(self, enabled=False):
+        """Enable or disable the log forwarding."""
+        if self.loki_endpoints:
+            layer_config = {"log-targets": self._build_log_targets(self.loki_endpoints, enabled)}
+            layer = Layer(layer_config)  # pyright: ignore
+
+            for container_name, container in self._charm.unit.containers.items():
+                container.add_layer(f"{container_name}-log-forwarding", layer, combine=True)
+        else:
+            logger.warning("No loki endpoints available")
+
+    @property
+    def _fetch_endpoints(self) -> List[str]:
+        """Fetch Loki Push API endpoints through relation data.
+
+        Returns:
+            A list of Loki Push API endpoint URLs, for instance:
+            [
+                "http://loki1:3100/loki/api/v1/push",
+                "http://loki2:3100/loki/api/v1/push",
+            ]
+        """
+        endpoints = []  # type: list
+
+        for relation in self._charm.model.relations[self._relation_name]:
+            for unit in relation.units:
+                if unit.app == self._charm.app:
+                    # This is a peer unit
+                    continue
+
+                endpoint = relation.data[unit].get("endpoint")
+                if endpoint:
+                    deserialized_endpoint = json.loads(endpoint)
+                    url = deserialized_endpoint.get("url")
+                    if url:
+                        endpoints.append(url)
+        endpoints.sort()
```

Comment:
why do we do this? test repeatability?

Reply:
I think so!

Reply:
That'd not be a good reason IMHO.
I think it's actually because otherwise the index in `_build_log_targets` might refer to different services every time, while we want the index to be persistent and always assigned to the same service.
This should be made clear in a comment otherwise it's not obvious why we're sorting that list. Also perhaps we shouldn't sort in this method but in `_build_log_targets`, where it's actually clear why we need the sorting.